### PR TITLE
exploratory uniswap agent to explore target grid price

### PR DESCRIPTION
**Pull request by msabvid**
- Remove uniswap agent type definition from core runner as it is passed as an argument
- Manually calibrate mu and sigma in DummyUniswapAgent in order for the GBM to explore a range of prices
- docstrings added

### CODE REVIEW

File: simulations/agents/liquidation_agent.py [hunk 1]
```diff
@@ -234,29 +329,76 @@ def record(self, env):
...
+        aave_oracle_abi: type,
         aave_oracle_address: bytes,
     ):
+        """
+        Initialise the Liquidator agent and create the corresponding
+        account in the EVM.
+
+        The agent stores the ABIs of the Aave contracts, the Uniswap contracts
+        and the token contracts that they will be interacting with.
+        ABIs are previously loaded using the function `:py:func:verbs.abi.load_abi`.
```
Comment (zombie-einstein) [inline simulations/agents/liquidation_agent.py:362]: Should be ` :py:fun:``verbs.abi.load_abi`` ` i.e the backticks should be after the `func` (should be single ticks, sorry could not work out how to show them)

File: simulations/agents/liquidation_agent.py [hunk 2]
```diff
@@ -290,6 +432,24 @@ def __init__(
...
+            Simulation environment.
+        liquidation_address: bytes
+            Liquidation address for which the Liquidator calculates the profitability
+            of the liquidation.
+        amount: int
+            Amount to be liquidated
+
+        Returns
+        -------
+        is_profitable: bool
```
Comment (zombie-einstein) [inline simulations/agents/liquidation_agent.py:451]: Just return type not name

File: simulations/agents/liquidation_agent.py [hunk 3]
```diff
@@ -301,7 +461,37 @@ def accountability(self, env, liquidation_address, amount: int) -> bool:
...
+            to ensure determinism of the simulation.
+        env: verbs.types.Env
+            Network/EVM that the simulation interacts with.
+
+        Returns
+        -------
+        list[Transaction]
+            List of transactions to be processed in the next block
+            of the simulation. This can be an empty list if the
+            agent is not submitting any transacti
```
Comment (zombie-einstein) [inline simulations/agents/liquidation_agent.py:493]: unfinished line

File: simulations/agents/uniswap_agent.py [hunk 4]
```diff
@@ -35,23 +35,48 @@ def tick_from_price(sqrt_price_x96: int, uniswap_fee: int) -> typing.Tuple[int,
...
+    Get price from tick
+
+    Parameters
+    ----------
+    tick: int
+        Lower tick of input price
+
+    Returns
+    -------
+    sqrt_price_x96: int
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:48]: Just type not name

File: simulations/agents/uniswap_agent.py [hunk 5]
```diff
@@ -35,23 +35,48 @@ def tick_from_price(sqrt_price_x96: int, uniswap_fee: int) -> typing.Tuple[int,
...
 class Gbm:
     """
-    Geometric brownian motion modelling the price of two tokens in USD
+    Geometric brownian motion modelling the price of two tokens
 
-    Notes
-    -----
-
-    We assume that token B is some stablecoin so its price remains constant.
+    .. note::
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:59]: You can just use "Notes" section then Sphinx will format it

File: simulations/agents/uniswap_agent.py [hunk 6]
```diff
@@ -65,8 +90,21 @@ def update(self, rng: np.random.Generator, price_impact: float):
 
         Update GBM price using:
 
-        - P^a_{t+dt} = P^a_t * exp((mu-0.5*sigma^2)dt + sigma * (W_{t+dt} - W_{t}))
-        - P^b is constant
+        * :math:`P^a_{t+dt} = P^a_t * exp((\\mu-0.5*\\sigma^2)dt + \sigma * (W_{t+dt} - W_{t}))`  # noqa E501
+        * :math:`P^{a, impact}_{t+dt} = P^a_{t+dt} + price_impact`
+        * :math:`P^b` is constant
+
+        .. note::
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:97]: Can use "Notes" section and sphinx will format it

File: simulations/agents/uniswap_agent.py [hunk 7]
```diff
@@ -79,11 +117,35 @@ def update(self, rng: np.random.Generator, price_impact: float):
         self.token_a_price = new_price_a
         self.token_a_price_with_impact = new_price_a_w_impact
 
-    def get_sqrt_price_token_a_x96(self):
+    def get_sqrt_price_token_a_x96(self) -> float:
+        """
+        Get price of token A in terms of token B
+
+        .. note::
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:124]: Can use "Notes" section

File: simulations/agents/uniswap_agent.py [hunk 8]
```diff
@@ -79,11 +117,35 @@ def update(self, rng: np.random.Generator, price_impact: float):
...
+            We return the square root of the price times 2**96 for a fair comparison
+            with the price values returnd by the Uniswap contract.
+
+        Returns
+        -------
+        sqrt_price_token_a_x96: float
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:130]: Just return type not name

File: simulations/agents/uniswap_agent.py [hunk 9]
```diff
@@ -79,11 +117,35 @@ def update(self, rng: np.random.Generator, price_impact: float):
...
+            Square root of the price of token A in terms of token B times 2**96
+        """
         price = self.token_a_price_with_impact / self.token_b_price
         return np.sqrt(price) * 2**96
 
     def get_sqrt_price_token_b_x96(self):
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:136]: Return type annotation

File: simulations/agents/uniswap_agent.py [hunk 10]
```diff
@@ -79,11 +117,35 @@ def update(self, rng: np.random.Generator, price_impact: float):
...
+        """
+        Get price of token B in terms of token A
+
+        .. note::
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:140]: Can use "Notes" section

File: simulations/agents/uniswap_agent.py [hunk 11]
```diff
@@ -79,11 +117,35 @@ def update(self, rng: np.random.Generator, price_impact: float):
...
+            We return the square root of the price times 2**96 for a fair comparison
+            with the price values returnd by the Uniswap contract.
+
+        Returns
+        -------
+        sqrt_price_token_a_x96: float
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:146]: return type only

File: simulations/agents/uniswap_agent.py [hunk 12]
```diff
@@ -112,6 +174,39 @@ def __init__(
         # token B is considered to be less risky / stablecoin
         token_b_address: bytes,
     ):
+        """
+        Initialise the Uniswap agent and create the corresponding
+        account in the EVM.
+
+        The agent stores the ABIs of the Uniswap contracts
+        and the token contracts that they will be interacting with.
+        ABIs are previously loaded using the function `:py:func:verbs.abi.load_abi`.
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:183]: Back quote should come after `:func:`

File: simulations/agents/uniswap_agent.py [hunk 13]
```diff
@@ -136,7 +231,21 @@ def __init__(
 
     def get_sqrt_price_x96_uniswap(self, env) -> int:
         """get sqrt price from uniswap pool.
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:233]: Should be on next line after quotes

File: simulations/agents/uniswap_agent.py [hunk 14]
```diff
@@ -136,7 +231,21 @@ def __init__(
...
+
         Uniswap returns price of token0 in terms of token1
+
+        .. note::
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:237]: Can use "Notes"

File: simulations/agents/uniswap_agent.py [hunk 15]
```diff
@@ -136,7 +231,21 @@ def __init__(
...
+            Uniswap sorts of token0 and token1 by their addresses.
+
+        Parameters
+        ----------
+        env: verbs.types.Env
+            Simulation environment
+
+        Returns
+        -------
+        sqrt_price_uniswap_x96: int
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:247]: Type only

File: simulations/agents/uniswap_agent.py [hunk 16]
```diff
@@ -163,10 +272,30 @@ def get_swap_size_to_increase_uniswap_price(
         numeraire (in our case the debt asset), and P is the price of the
         collateral in terms of the numeraire.
 
-        References
-        ----------
+        If there is a tick range and `exact=True`, the agent performs
+        an iterative calculation to find the right trade.
+
+        .. note::
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:278]: Can use "Notes"

File: simulations/agents/uniswap_agent.py [hunk 17]
```diff
@@ -163,10 +272,30 @@ def get_swap_size_to_increase_uniswap_price(
...
+            Sqrt of current uniswap price times 2**96
+        liquidity: int
+            Liquidity of Uniswap in the current tick range
+        exact: bool
+            Boolean indicating whether to perform the iterative calculation
+            to find the right trade.
+
+        Returns
+        -------
+        tx: verbs.types.Transaction
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:297]: Just type

File: simulations/agents/uniswap_agent.py [hunk 18]
```diff
@@ -242,6 +371,31 @@ def get_swap_size_to_decrease_uniswap_price(
         :math:`L = \\frac{\\Delta y}{\\Delta \\sqrt{P}}` where y is
         the numeraire (in our case the debt asset), and P is the price
         of the collateral in terms of the numeraire.
+
+        If there is a tick range and `exact=True`, the agent performs
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:375]: Rst code needs double back quotes

File: simulations/agents/uniswap_agent.py [hunk 19]
```diff
@@ -242,6 +371,31 @@ def get_swap_size_to_decrease_uniswap_price(
...
+        an iterative calculation to find the right trade.
+
+        .. note::
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:378]: Can use notes or References section

File: simulations/agents/uniswap_agent.py [hunk 20]
```diff
@@ -242,6 +371,31 @@ def get_swap_size_to_decrease_uniswap_price(
...
+            Sqrt of current uniswap price times 2**96
+        liquidity: int
+            Liquidity of Uniswap in the current tick range
+        exact: bool
+            Boolean indicating whether to perform the iterative calculation
+            to find the right trade.
+
+        Returns
+        -------
+        tx: verbs.types.Transaction
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:397]: Just return type

File: simulations/agents/uniswap_agent.py [hunk 21]
```diff
@@ -328,6 +480,48 @@ def __init__(
         sigma: float,
         dt: float,
     ):
+        """
+        Initialise the Uniswap agent and create the corresponding
+        account in the EVM.
+
+        The agent stores the ABIs of the Uniswap contracts
+        and the token contracts that they will be interacting with.
+        ABIs are previously loaded using the function `:py:func:verbs.abi.load_abi`.
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:489]: Back quote after `:verbs:`

File: simulations/agents/uniswap_agent.py [hunk 22]
```diff
@@ -429,9 +647,26 @@ def update(self, rng: np.random.Generator, env):
...
+        of the simulation.
+
+        Parameters
+        ----------
+        env: verbs.types.Env
+            Network/EVM that the simulation interacts with.
+
+        Returns
+        -------
+        price_uniswap: float
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:665]: Returns tuple type

File: simulations/agents/uniswap_agent.py [hunk 23]
```diff
@@ -453,10 +688,20 @@ def record(self, env):
...
+        on the external exchange.
+
+        Parameters
+        ----------
+        env: verbs.types.Env
+            Network/EVM that the simulation interacts with.
+
+        Returns
+        -------
+        impact: float
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:703]: Just return type

File: simulations/agents/uniswap_agent.py [hunk 24]
```diff
@@ -496,6 +741,55 @@ def __init__(
...
+        account in the EVM.
+
+        The agent stores the ABIs of the Uniswap contracts
+        and the token contracts that they will be interacting with.
+        ABIs are previously loaded using the function `:py:func:verbs.abi.load_abi`.
+
+        The agent also has access to an external market, modelled by a Gbm,
+        that is set as an attribute of the agents.
+
+        .. note::
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:755]: Can use "Notes"

File: simulations/agents/uniswap_agent.py [hunk 25]
```diff
@@ -496,6 +741,55 @@ def __init__(
         dt: float,
         sim_n_steps: int,
     ):
+        """
+        Initialise the Uniswap agent and create the corresponding
+        account in the EVM.
+
+        The agent stores the ABIs of the Uniswap contracts
+        and the token contracts that they will be interacting with.
+        ABIs are previously loaded using the function `:py:func:verbs.abi.load_abi`.
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:750]: Tick after `:verbs:`

File: simulations/agents/uniswap_agent.py [hunk 26]
```diff
@@ -511,19 +805,51 @@ def __init__(
             token_a_address=token_a_address,
             token_b_address=token_b_address,
             mu=mu,
-            sigma=sigma,
+            sigma=0,
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:808]: Did you want to set this?

File: simulations/agents/uniswap_agent.py [hunk 27]
```diff
@@ -511,19 +805,51 @@ def __init__(
...
+            to ensure determinism of the simulation.
+        env: verbs.types.Env
+            Network/EVM that the simulation interacts with.
+
+        Returns
+        -------
+        list[Transaction]
+            List of transactions to be processed in the next block
+            of the simulation. This can be an empty list if the
+            agent is not submitting any transacti
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:848]: Unfinished line

File: simulations/uniswap/plotting.py [hunk 28]
```diff
@@ -10,7 +10,19 @@
...
+
+    Will generate the following plots:
+    * Uniswap price of token0 in terms of token1
+    * External market price of token0 in terms of token1
+
+    Plots are saved in results/sim_uniswap_gbm
+
+    Parameters
+    ----------
+    results: list[list[tuple[int, int]]]
```
Comment (zombie-einstein) [inline simulations/uniswap/plotting.py:23]: Just return type

File: simulations/uniswap/plotting.py [hunk 29]
```diff
@@ -21,6 +33,6 @@ def plot_results(results: List[List[Tuple[int, int]]]):
 
     fig, ax = plt.subplots(figsize=(6, 3))
     ax.plot(prices[:, 0], label="Uniswap price")
-    ax.plot(prices[:, 1], label="External market price")
+    ax.plot(prices[:, 1], "--", label="External market price")
```
Comment (zombie-einstein) [inline simulations/uniswap/plotting.py:36]: What does this do (for my own knowledge)

**Review by zombie-einstein — COMMENTED** [review]
A few style comments

File: simulations/uniswap/plotting.py (hunk 29 above)
Comment (msabvid) [reply simulations/uniswap/plotting.py:36]: plots a dashed line

File: simulations/agents/uniswap_agent.py (hunk 26 above)
Comment (msabvid) [reply simulations/agents/uniswap_agent.py:815]: Yes, I'm making the exploration deterministic now, so no sigma.

**Pull request merged**